spec/openai/client/client_spec.rb

```diff
@@ -25,6 +25,88 @@
     end
   end
 
+  describe ".to_json_stream" do
+    context "with a proc" do
+      let(:user_proc) { proc { |x| x } }
+      let(:stream)    { OpenAI::Client.send(:to_json_stream, user_proc: user_proc) }
+
+      it "returns a proc" do
+        expect(stream).to be_a(Proc)
+      end
+
+      context "when called with a string containing a single JSON object" do
+        it "calls the user proc with the data parsed as JSON" do
+          expect(user_proc).to receive(:call).with(JSON.parse('{"foo": "bar"}'))
+          stream.call('data: { "foo": "bar" }')
+        end
+      end
+
+      context "when called with string containing more than one JSON object" do
+        it "calls the user proc for each data parsed as JSON" do
+          expect(user_proc).to receive(:call).with(JSON.parse('{"foo": "bar"}'))
+          expect(user_proc).to receive(:call).with(JSON.parse('{"baz": "qud"}'))
+
+          stream.call(<<-JSON)
+            data: { "foo": "bar" }
+
+            data: { "baz": "qud" }
+
+            data: [DONE]
+
+          JSON
+        end
+      end
+
+      context "when called with a string that does not even resemble a JSON object" do
+        let(:bad_examples) { ["", "foo", "data: ", "data: foo"] }
+
+        it "does not call the user proc" do
+          bad_examples.each do |chunk|
+            expect(user_proc).to_not receive(:call)
+            stream.call(chunk)
+          end
+        end
+      end
+
+      context "when called with a string containing that looks like a JSON object but is invalid" do
+        let(:chunk) do
+          <<-CHUNK
+            data: { "foo": "bar" }
+            data: { BAD ]:-> JSON }
+          CHUNK
+        end
+
+        it "does not raise an error" do
+          expect(user_proc).to receive(:call).with(JSON.parse('{"foo": "bar"}'))
+
+          expect do
+            stream.call(chunk)
+          end.to_not raise_error(JSON::ParserError)
+        end
+      end
+
+      context "when called with a string containing an error" do
+        let(:chunk) do
+          <<-CHUNK
+            data: { "foo": "bar" }
+            error: { message: "A bad thing has happened!" }
```

Comment:
I think this is fine in the spec context to make sure it filters out anything that doesn't look like valid data, but I think errors would also be sent as `data: {}` though, see https://github.com/alexrudall/ruby-openai/pull/234#discussion_r1176889900

which is not problem necessarily as it allows callers to handle those errors in the same callback

Reply:
Yes. This is an open question. There are two kinds of errors here that seem like they will come up often. Parsing errors and just bad data seems less likely but possible. Then there are the "good" errors coming from OpenAI where they return an `error: ...` message. What should happen to those "good" error messages? It seems like it would be a good idea to yield those errors to the user_proc for now as JSON so the user can do whatever they would like with them (skip, try again, etc.). I'm not sure if skipped, the next chunk of valid data would pick back up where it left off or not.

Reply:
@fwitzke I decided to send the OpenAI errors to the user proc.